src/settings/login/secondfactor/SecondFactorEditModel.ts

```diff
@@ -0,0 +1,270 @@
+			} else {
+				totpQRCodeSvg = null
+			}
+
+			return {
+				qrCodeSvg: totpQRCodeSvg,
+				url,
+			}
+		})
+
+		this.otpInfo.getAsync().then(() => this.updateViewCallback())
+	}
+
+	abort() {
+		// noinspection JSIgnoredPromiseFromCall
+		this.webauthnClient.abortCurrentOperation()
+	}
+
+	validator() {
+		return () => this.updateNameValidation()
+	}
+
+	validationMessage(): TranslationKey | null {
+		return this.nameValidationStatus === NameValidationStatus.Valid
+			? null
+			: "textTooLong_msg"
+	}
+
+	getFactorTypesOptions() {
+		const options: Array<SelectorItem<FactorTypesEnum>> = []
+		options.push({
+			name: lang.get("totpAuthenticator_label"),
+			value: FactorTypes.TOTP,
+		})
+
+		if (this.webauthnSupported) {
+			options.push({
+				name: lang.get("u2fSecurityKey_label"),
+				value: FactorTypes.WEBAUTHN,
+			})
+		}
+		return options
+	}
+
+	onTypeSelected(newValue: FactorTypesEnum) {
+		this.selectedType = newValue
+		this.verificationStatus = newValue === FactorTypes.WEBAUTHN
+			? VerificationStatus.Initial
+			: VerificationStatus.Progress
+
+		this.updateNameValidation()
+		this.setDefaultNameIfNeeded()
+
+		if (newValue !== FactorTypes.WEBAUTHN) {
+			// noinspection JSIgnoredPromiseFromCall
+			this.webauthnClient.abortCurrentOperation()
+		}
+	}
+
+	onNameChange(newValue: string): void {
+		this.name = newValue
+		this.updateNameValidation()
+	}
+
+	/**
+	 * call when the validation code for setting up TOTP changes
+	 */
+	async onTotpValueChange(newValue: string) {
+		this.totpCode = newValue
+		let cleanedValue = newValue.replace(/ /g, "")
+
+		if (cleanedValue.length === 6) {
+			const expectedCode = Number(cleanedValue)
+			this.verificationStatus = await this.tryCodes(expectedCode, this.totpKeys.key)
+		} else {
+			this.verificationStatus = VerificationStatus.Progress
+		}
+	}
+
+	/**
+	 * validates the input and makes the server calls to actually create a second factor
+	 * if the validation passes.
+	 */
+	async save(callback: (user: User) => void): Promise<void> {
+		this.setDefaultNameIfNeeded()
+		if (this.selectedType === FactorTypes.WEBAUTHN) {
+			// Prevent starting in parallel
+			if (this.verificationStatus === VerificationStatus.Progress) {
+				return
+			}
+
+			try {
+				this.u2fRegistrationData = await this.webauthnClient.register(this.user._id, this.name)
+				this.verificationStatus = VerificationStatus.Success
+			} catch (e) {
+				console.log("Webauthn registration failed: ", e)
+				this.u2fRegistrationData = null
+				this.verificationStatus = VerificationStatus.Failed
+			}
+		}
+
+		this.updateViewCallback()
+		const sf = createSecondFactor({
+			_ownerGroup: this.user._ownerGroup,
+			name: this.name,
+		})
+
+		if (this.selectedType === FactorTypes.WEBAUTHN) {
+			sf.type = SecondFactorType.webauthn
+
+			if (this.verificationStatus !== VerificationStatus.Success) {
+				// noinspection ES6MissingAwait
+				Dialog.message("unrecognizedU2fDevice_msg")
```

Comment:
dialogs from models is not not great but if it's too hard to move around we can keep it. I'd just throw `UserError` and catch it on the client